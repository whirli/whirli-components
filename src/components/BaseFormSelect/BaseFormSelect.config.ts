// @ts-ignore
import { PropType } from '@composition';
import * as PropHelpers from '@whirli-components/helpers/props';
import { ComponentStyles, ComponentProps } from '@whirli-components/@types/components';

// Local config
import config from '@whirli-local/components/BaseFormSelect/BaseFormSelect.config';

// Shared Config
import { SharedConfigProps } from '@whirli-components/components/BaseFormInput/BaseFormInput.config';

// Package config
import { PropKeys, PropValues } from './BaseFormSelect.constants';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
};

export const ConfigProps: ComponentProps = {
  [PropKeys.VALUE]: {
    required: true,
    type: [String, Number] as PropType<string | number>,
  },
  [PropKeys.DROPDOWN_MESSAGE]: {
    required: false,
    type: String as PropType<string>,
    default: 'Please select an option',
  },
  [PropKeys.SELECTED]: {
    required: false,
    type: String as PropType<string>,
    default: '',
  },
  ...SharedConfigProps,
  ...config.props,
};
