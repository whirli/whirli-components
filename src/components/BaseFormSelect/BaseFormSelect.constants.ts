export const PropKeys: Record<string, string> = {
  MODAL_VALUE: 'modelValue',
  HAS_DROPDOWN_MESSAGE: 'hasDropdownMessage',
  DROPDOWN_MESSAGE: 'dropdownMessage',
  SELECTED: 'selected',
};

export const PropValues: Record<string, string[]> = {};
