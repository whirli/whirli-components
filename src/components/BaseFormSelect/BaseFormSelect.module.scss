@import '../../styles/Breakpoints.module.scss';

// default styles

// breakpoint styles
$packageBreakpointStyles: ();