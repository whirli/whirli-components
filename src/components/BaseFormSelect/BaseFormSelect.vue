<template>
  <BaseFormGroupLayoutDefault v-bind="$props">
    <select :id="props.name" :name="props.name" :class="classes">
      <slot />
    </select>
  </BaseFormGroupLayoutDefault>
</template>

<script setup lang="ts">
// Styles
// @ts-ignore
import styles from '@whirli-local/components/BaseFormInput/BaseFormInput.module.scss?module';

// Data
import {
  ConfigStyles,
  ConfigProps,
} from '@whirli-components/components/BaseFormGroup/BaseFormGroupLayout.config';

// Types
import { ComponentStyles as ComponentStylesInterface } from '@whirli-components/@types/components';
import { Props } from '@whirli-components/@types/props';

const ComponentStyles: ComponentStylesInterface = ConfigStyles;

// @ts-ignore
const props: Props = defineProps(ConfigProps);

// Components
import BaseFormGroupLayoutDefault from '@whirli-components/components/BaseFormGroup/BaseFormGroupLayoutDefault.vue';

// Classes
import useClasses from '@whirli-components/@use/class';
const { makeClasses } = useClasses();
const classes = [styles.input, styles.select, ...makeClasses(ComponentStyles, props, styles)];
</script>
