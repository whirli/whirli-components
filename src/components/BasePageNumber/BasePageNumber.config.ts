// @ts-ignore
import { PropType } from 'vue';
import * as PropHelpers from '../../helpers/props';
import { ComponentStyles, ComponentProps } from '../../@types/components';

// Local config
import config from '@whirli-local/components/BasePageNumber/BasePageNumber.config';

// Package config
import { PropKeys } from './BasePageNumber.constants';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
};

export const ConfigProps: ComponentProps = {
  [PropKeys.PAGE_NUMBER]: {
    required: true,
    type: Number as PropType<number>,
  },
  ...config.props,
};
