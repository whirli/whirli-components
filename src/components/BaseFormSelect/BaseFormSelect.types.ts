import { PropKeys, PropValues } from './BaseFormSelect.constants';

// const propColor: string[]  = PropValues[PropKeys.COLOR];

// export type PropColor = typeof propColor[number];
