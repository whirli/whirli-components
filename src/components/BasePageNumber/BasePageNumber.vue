<template>
  <component
    :is="tag"
    :href="pageUrl"
    :class="[defaultClasses, activeClasses, disabledClasses]"
    @click.prevent="$emit('page-number:click', props.pageNumber)"
  >
    {{ props.pageNumber }}
  </component>
</template>

<script setup lang="ts">
// Vue
import { defineProps, defineEmits, computed } from 'vue';
import type { ComputedRef } from 'vue';
import { useRoute } from 'vue-router';

defineEmits(['page-number:click']);
const Route = useRoute();

// Styles
import styles from '@whirli-local/components/BasePageNumber/BasePageNumber.module.scss';

// Data
import { ConfigStyles, ConfigProps } from './BasePageNumber.config';
import { PropKeys } from '../BasePageNumber/BasePageNumber.constants';

// Types
import { ComponentStyles as ComponentStylesInterface } from '../../@types/components';
import { Props } from '../../@types/props';

const ComponentStyles: ComponentStylesInterface = ConfigStyles;

const props: Props = defineProps(ConfigProps);

const pageUrl: ComputedRef<string> = computed(() => `${Route.path}?page=${props.pageNumber}`);
const isActivePage: ComputedRef<boolean> = computed(
  () =>
    (Route.query.page && +Route.query.page === props.pageNumber) ||
    (!Route.query.page && props.pageNumber === 1)
);

const tag: ComputedRef<string> = computed(() => (isActivePage.value ? 'div' : 'a'));

// Classes
import useClasses from '../../@use/class';
const { makeClasses } = useClasses();
const defaultClasses = [styles['page-number'], ...makeClasses(ComponentStyles, props, styles)];
const activeClasses: ComputedRef<string> = computed(() => (isActivePage.value ? styles['is-active'] : ''));
const disabledClasses: ComputedRef<string> = computed(() =>
  props[PropKeys.IS_DISABLED] ? styles['is-disabled'] : ''
);
</script>
