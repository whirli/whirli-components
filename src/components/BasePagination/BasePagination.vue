<template>
  <nav :class="wrapperClasses">
    <BaseGrid>
      <BaseRow justify="between" align="center">
        <BaseIconButton
          tag="a"
          icon="ArrowCenter"
          :href="previousPageUrl"
          @click.prevent="goToPreviousPage()"
          backdrop="circle"
          rotate="90"
          ref="previousButton"
          :is-disabled="isFirstPage"
          aria-label="Previous page"
        />
        <BaseList direction="row" class="from-tablet">
          <BaseListItem v-for="(option, index) in getPageNumbers" :key="`page-number-${index}`">
            <BaseText v-if="option.isPlaceholder">{{ props.placeholder }}</BaseText>
            <BasePageNumber
              v-else-if="option"
              :is-disabled="props.busy"
              :page-number="option.number"
              @page-number:click="goToPage(option.number ?? 1)"
            />
          </BaseListItem>
        </BaseList>
        <BaseText ref="pageCount" class="until-tablet">
          Page {{ currentPage }} of {{ props.lastPage }}
        </BaseText>
        <BaseIconButton
          tag="a"
          icon="ArrowCenter"
          :href="nextPageUrl"
          @click.prevent="goToNextPage()"
          backdrop="circle"
          rotate="270"
          ref="nextButton"
          :is-disabled="isLastPage"
          aria-label="Next page"
        />
      </BaseRow>
      <BaseRow class="until-tablet">
        <BaseFormSelect name="pagination" label="Jump to" @change="goToPage($event.target.value)">
          <option
            v-for="(number, index) in props.lastPage"
            :key="index"
            :value="number"
            :selected="number === currentPage"
          >
            {{ number }}
          </option>
        </BaseFormSelect>
      </BaseRow>
    </BaseGrid>
  </nav>
</template>

<script setup lang="ts">
// Vue
import { computed, defineProps, defineEmits } from 'vue';
import type { ComputedRef } from 'vue';
import { useRoute } from 'vue-router';

const Route = useRoute();
const Emit = defineEmits(['pagination:go-to', 'pagination:next', 'pagination:previous']);

// Styles
import styles from '@whirli-local/components/BasePagination/BasePagination.module.scss';

// Data
import { ConfigStyles, ConfigProps } from './BasePagination.config';

// Types
import { ComponentStyles as ComponentStylesInterface } from '../../@types/components';
import { Props } from '../../@types/props';
import type { PaginationOption } from '../../@types/pagination';

// Components
import BaseGrid from '../BaseGrid/BaseGrid.vue';
import BaseRow from '../BaseRow/BaseRow.vue';
import BasePageNumber from '../BasePageNumber/BasePageNumber.vue';
import BaseIconButton from '../BaseIconButton/BaseIconButton.vue';
import BaseList from '../BaseList/BaseList.vue';
import BaseListItem from '../BaseListItem/BaseListItem.vue';
import BaseText from '../BaseText/BaseText.vue';
import BaseFormSelect from '../BaseFormSelect/BaseFormSelect.vue';

const ComponentStyles: ComponentStylesInterface = ConfigStyles;

const props: Props = defineProps(ConfigProps);

const firstPage = 1;
const currentPage: ComputedRef<number> = computed(() => (Route.query.page ? +Route.query.page : 1));

const isFirstPage: ComputedRef<boolean> = computed(() => currentPage.value === firstPage);
const isLastPage: ComputedRef<boolean> = computed(() => currentPage.value === props.lastPage);

// Classes
import useClasses from '../../@use/class';
const { makeClasses } = useClasses();
const wrapperClasses = [styles.pagination, ...makeClasses(ComponentStyles, props, styles)];

function isCurrentPage(number: number): boolean {
  return number === currentPage.value;
}

function goToPage(page: number) {
  if (page < firstPage || page > props.lastPage) return;
  Emit('pagination:go-to', page);
}

function goToNextPage() {
  if (currentPage.value === props.lastPage) return;
  Emit('pagination:next');
}

function goToPreviousPage() {
  if (currentPage.value === firstPage) return;
  Emit('pagination:previous');
}

const previousPageUrl: ComputedRef<string | boolean> = computed(() => {
  const previousPageNumber: number = currentPage.value - 1;
  return previousPageNumber < 1 ? false : `${Route.path}?page=${previousPageNumber}`;
});

const nextPageUrl: ComputedRef<string | boolean> = computed(() => {
  const nextPageNumber: number = currentPage.value + 1;
  return nextPageNumber > props.lastPage ? false : `${Route.path}?page=${nextPageNumber}`;
});

const getPageNumbers: ComputedRef<PaginationOption[]> = computed(() => {
  const placeholderOption: PaginationOption = { isPlaceholder: true };

  const lowerBound = Math.max(2, currentPage.value - (props.delta as number));
  const upperBound = Math.min((props.lastPage as number) - 1, currentPage.value + (props.delta as number));

  const range = [];
  for (let pageNumber = lowerBound; pageNumber <= upperBound; pageNumber += 1) {
    range.push({ number: pageNumber, isPlaceholder: false, isCurrentPage: isCurrentPage(pageNumber) });
  }

  if (currentPage.value - (props.delta as number) > 2) range.unshift(placeholderOption);

  if (currentPage.value + (props.delta as number) < (props.lastPage as number) - 1)
    range.push(placeholderOption);

  range.unshift({ number: 1, isPlaceholder: false, isCurrentPage: isCurrentPage(1) });

  if (props.lastPage !== 1) {
    range.push({
      number: props.lastPage as number,
      isPlaceholder: false,
      isCurrentPage: isCurrentPage(props.lastPage as number),
    });
  }

  return range;
});
</script>
