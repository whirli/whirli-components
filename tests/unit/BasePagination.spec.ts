import { mount } from '@vue/test-utils';
import BasePagination from '@/components/BasePagination/BasePagination.vue';
import BasePageNumber from '@/components/BasePageNumber/BasePageNumber.vue';
import BaseFormSelect from '@/components/BaseFormSelect/BaseFormSelect.vue';

let wrapper: any;
const firstPage = 1;
const lastPage = 30;
const pageNumber = '3';

jest.mock('vue-router', () => ({
  useRoute: () => ({
    path: '/browse',
    query: {
      page: '3',
    },
  }),
}));

describe('BasePagination.vue', () => {
  beforeEach(() => {
    wrapper = mount(BasePagination, {
      props: {
        firstPage,
        lastPage,
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('emmits next on prev button click', async () => {
    wrapper.findComponent({ ref: 'previousButton' }).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('pagination:previous'));
  });

  it('emmits next on next button click', async () => {
    wrapper.findComponent({ ref: 'nextButton' }).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('pagination:next'));
  });

  it('emits page select on PageNumber change', async () => {
    wrapper.findComponent(BasePageNumber).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('pagination:go-to'));
  });

  it('emits page select on BaseSelect change', async () => {
    wrapper.findComponent(BaseFormSelect).trigger('change');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('pagination:go-to'));
  });

  it('has the correct number of page select options', async () => {
    expect(wrapper.findAll('option').length).toBe(30);
  });

  it('has the correct pageCount text', async () => {
    expect(wrapper.findComponent({ ref: 'pageCount' }).text()).toBe(`Page ${pageNumber} of ${lastPage}`);
  });
});
