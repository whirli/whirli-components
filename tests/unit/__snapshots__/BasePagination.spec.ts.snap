// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasePagination.vue renders the component 1`] = `
<nav
  class=""
  firstpage="1"
>
  <div
    class=""
  >
    
    <div
      class=""
    >
      
      
      <a
        aria-label="Previous page"
        backdrop="circle"
        class=""
        href="/browse?page=2"
        icon="ArrowCenter"
        rotate="90"
      >
        <svg
          aria-label="Previous page"
          class=""
          href="/browse?page=2"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 11.6l8 8 8-8-1.4-1.3-5.6 5.6V4.5h-2v11.4l-5.5-5.6L4 11.6z"
          />
        </svg>
      </a>
      <ul
        class="from-tablet"
      >
        
        
        <li
          class=""
        >
          <div
            class=""
          >
            
            <a
              class=""
              href="/browse?page=1"
            >
              1
            </a>
            
          </div>
        </li>
        <li
          class=""
        >
          <div
            class=""
          >
            
            <a
              class=""
              href="/browse?page=2"
            >
              2
            </a>
            
          </div>
        </li>
        <li
          class=""
        >
          <div
            class=""
          >
            
            <div
              class=""
              href="/browse?page=3"
            >
              3
            </div>
            
          </div>
        </li>
        <li
          class=""
        >
          <div
            class=""
          >
            
            <a
              class=""
              href="/browse?page=4"
            >
              4
            </a>
            
          </div>
        </li>
        <li
          class=""
        >
          <div
            class=""
          >
            
            <a
              class=""
              href="/browse?page=5"
            >
              5
            </a>
            
          </div>
        </li>
        <li
          class=""
        >
          <div
            class=""
          >
            
            <a
              class=""
              href="/browse?page=6"
            >
              6
            </a>
            
          </div>
        </li>
        <li
          class=""
        >
          <div
            class=""
          >
            
            <span
              class=""
            >
              
              ...
              
            </span>
            
          </div>
        </li>
        <li
          class=""
        >
          <div
            class=""
          >
            
            <a
              class=""
              href="/browse?page=30"
            >
              30
            </a>
            
          </div>
        </li>
        
        
      </ul>
      <span
        class="until-tablet"
      >
        
         Page 3 of 30
        
      </span>
      <a
        aria-label="Next page"
        backdrop="circle"
        class=""
        href="/browse?page=4"
        icon="ArrowCenter"
        rotate="270"
      >
        <svg
          aria-label="Next page"
          class=""
          href="/browse?page=4"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 11.6l8 8 8-8-1.4-1.3-5.6 5.6V4.5h-2v11.4l-5.5-5.6L4 11.6z"
          />
        </svg>
      </a>
      
      
    </div>
    <div
      class="until-tablet"
    >
      
      
      <div
        class=""
      >
        
        <div
          class=""
        >
          
          <div
            class=""
          >
            
            
            <label
              class=""
              for="pagination"
            >
              
              Jump to
              
            </label>
            
            
          </div>
          <div
            class=""
          >
            
            
            
            <select
              class=""
              id="pagination"
              name="pagination"
            >
              
              
              <option
                value="1"
              >
                1
              </option>
              <option
                value="2"
              >
                2
              </option>
              <option
                value="3"
              >
                3
              </option>
              <option
                value="4"
              >
                4
              </option>
              <option
                value="5"
              >
                5
              </option>
              <option
                value="6"
              >
                6
              </option>
              <option
                value="7"
              >
                7
              </option>
              <option
                value="8"
              >
                8
              </option>
              <option
                value="9"
              >
                9
              </option>
              <option
                value="10"
              >
                10
              </option>
              <option
                value="11"
              >
                11
              </option>
              <option
                value="12"
              >
                12
              </option>
              <option
                value="13"
              >
                13
              </option>
              <option
                value="14"
              >
                14
              </option>
              <option
                value="15"
              >
                15
              </option>
              <option
                value="16"
              >
                16
              </option>
              <option
                value="17"
              >
                17
              </option>
              <option
                value="18"
              >
                18
              </option>
              <option
                value="19"
              >
                19
              </option>
              <option
                value="20"
              >
                20
              </option>
              <option
                value="21"
              >
                21
              </option>
              <option
                value="22"
              >
                22
              </option>
              <option
                value="23"
              >
                23
              </option>
              <option
                value="24"
              >
                24
              </option>
              <option
                value="25"
              >
                25
              </option>
              <option
                value="26"
              >
                26
              </option>
              <option
                value="27"
              >
                27
              </option>
              <option
                value="28"
              >
                28
              </option>
              <option
                value="29"
              >
                29
              </option>
              <option
                value="30"
              >
                30
              </option>
              
              
            </select>
            
            
            
          </div>
          <!--v-if-->
          
        </div>
        
      </div>
      
      
    </div>
    
  </div>
</nav>
`;
